### DISCUSSION TITLE
Strimzi topic operator creating automatic topics

### DISCUSSION THREAD

**Discussion by shreyasarani23** [strimzi/strimzi-kafka-operator#5358] [General]
Following is the yaml file for creating topics
```
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaTopic
metadata:
  name: ad-click
  labels:
    strimzi.io/cluster: my-kafka-cluster
spec:
  topicName: ad_click
  partitions: 3
  replicas: 3
  config:
    min.insync.replicas: 2
```
I want my topic to be named as ad_click. But I know that kubernetes doesn't support metadata.name to have underscore. So I have used spec.topicName and created the topic. But when I do ```kubectl get KafkaTopics -n kafka``` I see that the Topic Operator has created topics by appending a hash to the topic name. Even though I have used ```spec.topicName``` why it is creating the topics appending hash to the topic? And also these are not in ready state.

please help me to resolve this issue.
```
kubectl get KafkaTopics -n kafka
NAME                                                                                               CLUSTER            PARTITIONS   REPLICATION FACTOR   READY
ad-click                                                                                           my-kafka-cluster   3            3                    True
ad-click---8d8d553a97f8d2c936d8069b16346eb58e209634                                                my-kafka-cluster   3            3
add-reminder-icon                                                                                  my-kafka-cluster   2            3                    True
add-reminder-icon---a5517286a3495ccd16c5b588577ee0a64f666871                                       my-kafka-cluster   2            3
app-exit                                                                                           my-kafka-cluster   3            3                    True
app-exit---28aefee7fbc40e162373c31d2f6559fdacffa030                                                my-kafka-cluster   3            3
```
```
kubectl describe KafkaTopic ad-click---8d8d553a97f8d2c936d8069b16346eb58e209634  -n kafka
Name:         ad-click---8d8d553a97f8d2c936d8069b16346eb58e209634
Namespace:    kafka
Labels:       strimzi.io/cluster=my-kafka-cluster
Annotations:  <none>
API Version:  kafka.strimzi.io/v1beta2
Kind:         KafkaTopic
Metadata:
  Creation Timestamp:  2021-07-13T17:50:48Z
  Generation:          1
  Managed Fields:
    API Version:  kafka.strimzi.io/v1beta2
    Fields Type:  FieldsV1
    fieldsV1:
      f:status:
        f:conditions:
    Manager:         okhttp
    Operation:       Update
    Time:            2021-07-13T17:51:04Z
  Resource Version:  19939221
  Self Link:         /apis/kafka.strimzi.io/v1beta2/namespaces/kafka/kafkatopics/ad-click---8d8d553a97f8d2c936d8069b16346eb58e209634
  UID:               d8394fc8-c335-4b70-b114-cc6c82a4bbc1
Spec:
  Config:
    min.insync.replicas:  2
  Partitions:             3
  Replicas:               3
  Topic Name:             ad_click
Status:
  Conditions:
    Last Transition Time:  2021-07-13T17:51:03.987012Z
    Message:               The state store, topic-store, may have migrated to another instance.
    Reason:                InvalidStateStoreException
    Status:                True
    Type:                  NotReady
  Observed Generation:     1
  Topic Name:              ad_click
Events:                    <none>
```

**Comment by sknot-rh** [comment]
What version of Strimzi do you use? I tried your reproducer with the latest image. Everything works as expected and the Topic resource is ready. Also listing topics from the Kafka broker contains the correctly named topic.

```
[sknot@localhost strimzi-kafka-operator]$ oc get kt 
NAME                                                                                               CLUSTER      PARTITIONS   REPLICATION FACTOR   READY
ad-click                                                                                           my-cluster   3            3                    True
consumer-offsets---84e7a678d08f4bd226872e5cdd4eb527fadc1c6a                                        my-cluster   50           3                    True
strimzi-store-topic---effb8e3e057afce1ecf67c3f5d8e4e3ff177fc55                                     my-cluster   1            3                    True
strimzi-topic-operator-kstreams-topic-store-changelog---b75e702040b99be8a9263134de3507fc0cc4017b   my-cluster   1            1                    True
```
```
[sknot@localhost tmp]$ oc exec -ti my-cluster-kafka-0 -- bash bin/kafka-topics.sh --list --bootstrap-server localhost:9092
__consumer_offsets
__strimzi-topic-operator-kstreams-topic-store-changelog
__strimzi_store_topic
ad_click
```

**Reply by shreyasarani23** [reply]
Hi @sknot-rh even I am using latest version of strimzi i.e. 0.24.0. But in my case why the topic operator  created the topics appending hash to the topic name. FYI I am using AKS ( azure kubernetes service) I had stopped the cluster and started the cluster today. Did this operation trigger topic operator to create topics?

Any idea why topic operator is creating topics by appending hash?

**Reply by sknot-rh** [reply]
Appending a hash to the topic name is how we cope with "translation" between valid kafka topic names and valid kubernetes resource names. If you create a topic in kafka broker manually and give it some kube-invalid name, the hash is appended.
What does stopping the cluster mean? Did you remove any resources?

**Reply by shreyasarani23** [reply]
Stopping the AKS cluster means I am shutting down the AKS cluster. when I shutdown the cluster all my pods including kafka will be down.

**Reply by scholzj** [reply]
@shreyasarani23 This can happen in many situation. It could for example also be that while you are creating the topic as Kafka Topic, your client might be creating the topic on it own in the Kafka cluster - for example using auto-creation. That creates all kind of race conditions. The general recommendation is to either: 
* Use the topic names which are valid Kubernetes names (i.e. use `ad-click` instead of `ad_click` in the first place)
* Or use the full name including the hash as that creates deterministic naming without any conflicts

But why exactly you have two of them can be probably determined only from the full logs from when the topics were created.

PS: You are missing the internal Topic Operator topics in our listing. Not sure why, but they normally need to be there.

**Reply by shreyasarani23** [reply]
Attaching the topic operator logs
```
kubectl logs my-kafka-cluster-entity-operator-8499b4b74c-6w72f -c topic-operator -n kafka | grep ad-click
2021-07-26 09:04:10,79348 INFO  [vert.x-eventloop-thread-1] TopicOperator:576 - Reconciliation #583928(periodic kafka ad_click) KafkaTopic(kafka/ad-click---8d8d553a97f8d2c936d8069b16346eb58e209634): Reconciling topic ad_click, k8sTopic:nonnull, kafkaTopic:nonnull, privateTopic:nonnull
2021-07-26 09:04:10,79359 INFO  [vert.x-eventloop-thread-1] TopicOperator:743 - Reconciliation #583933(periodic kafka ad_click) KafkaTopic(kafka/ad-click---8d8d553a97f8d2c936d8069b16346eb58e209634): All three topics are identical
2021-07-26 09:04:10,79373 INFO  [vert.x-eventloop-thread-1] TopicOperator:1493 - Reconciliation #583934(periodic kafka ad_click) KafkaTopic(kafka/ad-click---8d8d553a97f8d2c936d8069b16346eb58e209634): Success reconciling KafkaTopic kafka/ad-click
```
FYI I have used ```auto.create.topics.enable: "false"``` in my kafka configuration. I doubt client will be able to create topics on its own. 
Can you infer anything from the logs?

I have the internal topics
```
kubectl get KafkaTopics -n kafka | grep consumer
consumer-offsets---84e7a678d08f4bd226872e5cdd4eb527fadc1c6a                                        my-kafka-cluster   50           3                    True

 kubectl get KafkaTopics -n kafka | grep strimzi
strimzi-store-topic---effb8e3e057afce1ecf67c3f5d8e4e3ff177fc55                                     my-kafka-cluster   1            3                    True
strimzi-topic-operator-kstreams-topic-store-changelog---b75e702040b99be8a9263134de3507fc0cc4017b   my-kafka-cluster   1            1                    True
```

**Reply by scholzj** [reply]
> Does stopping of the cluster has anything to do with internal topics missing?

From my experience, Kubernetes does not handle stopping and starting the cluster well in many cases. Things do not start always in the right order and that can cause all kind of issues. But I do not know if this is related. 

> And yes I am missing the internal topics don't know why they are not present.

I guess someone / something must have deleted the topics. Since they store the state of the operator, all kind of things can go wrong when they are deleted. You should maybe try to restart the topic operator and see if they get created.

> Attaching the topic operator logs

I'm afraid this is useless. You would need to provide the full log from the time the issue occurred (when the KafkaTopic CRs were created).

> FYI I have used auto.create.topics.enable: "false" in my kafka configuration. I doubt client will be able to create topics on its own.

`auto.create.topics.enable: "false"` is a good start. A lot of applications are creating their topics programatically. Not sure whether that is your case or not. But this happening due to the auto.creation is IMHO most common.

**Reply by shreyasarani23** [reply]
sorry I didn't check properly I have the internal topics
```
kubectl get KafkaTopics -n kafka | grep consumer
consumer-offsets---84e7a678d08f4bd226872e5cdd4eb527fadc1c6a                                        my-kafka-cluster   50           3                    True

 kubectl get KafkaTopics -n kafka | grep strimzi
strimzi-store-topic---effb8e3e057afce1ecf67c3f5d8e4e3ff177fc55                                     my-kafka-cluster   1            3                    True
strimzi-topic-operator-kstreams-topic-store-changelog---b75e702040b99be8a9263134de3507fc0cc4017b   my-kafka-cluster   1            1                    True
```

**Reply by shreyasarani23** [reply]
So what can I do now is there any solution ?
I have around 70-75 topics deleting each topic associated with hash manually  is a tedious task.
Any other work around?

**Reply by sknot-rh** [reply]
I do not tnink there is any workaround :( if the hsh-suffixed topics are created, that means the mapping between hashed-topic->kube resource already exists in `strimzi-store-topic`.
Maybe this could work. But I am not 100% sure it does the right thing. @alesj can you help there?
-  stopping TO
- removing all k8s Kafka Topic resources
- removing `strimzi-store-topic` from the kafka
- creating all your KafkaTopic resources
- starting TO

**Reply by shreyasarani23** [reply]
Hi just an update I was able to delete the topics associated with hash with the following command
```
kubectl get KafkaTopic -n kafka | grep -v True | awk '{print $1}' | xargs kubectl -n kafka delete KafkaTopic

```
-v ->  --invert-match        select non-matching lines
awk '{print $1}' -> prints only 1st column i.e. topic name appended with hash
xargs-> passes the output of awk command  to the kubectl command.

All the topics will be deleted one by one. But still not clear why these were created,

**Reply by sknot-rh** [reply]
I am glad you fixed your cluster. As jakub suggested the full log from the time the issue occurred would help us to understand what exactly happened. My guess is that TO recreated the k8s KafkaTopics from topics present in the brokers before they were created by you.
